blockchain/reactor_fsm.go

```diff
@@ -0,0 +1,430 @@
+					fsm.logger.Error("send peer error for", "peer", second.peer.id)
+					fsm.toBcR.sendPeerError(data.err, second.peer.id)
+					// Remove the first two blocks. This will also remove the peers
+					fsm.pool.invalidateFirstTwoBlocks(data.err)
+				} else {
+					fsm.pool.processedCurrentHeightBlock()
+					// Since we advanced one block reset the state timer
+					fsm.resetStateTimer()
+				}
+
+				// Both cases above may result in achieving maximum height.
+				if fsm.pool.reachedMaxHeight() {
+					return finished, nil
+				}
+
+				return waitForBlock, data.err
+
+			case peerRemoveEv:
+				// This event is sent by the switch to remove disconnected and errored peers.
+				fsm.pool.removePeer(data.peerId, data.err)
+				if len(fsm.pool.peers) == 0 {
+					return waitForPeer, nil
+				}
+				if fsm.pool.reachedMaxHeight() {
+					return finished, nil
+				}
+				return waitForBlock, nil
+
+			case makeRequestsEv:
+				fsm.makeNextRequests(data.maxNumRequests)
+				return waitForBlock, nil
+
+			case stateTimeoutEv:
+				if data.stateName != "waitForBlock" {
+					fsm.logger.Error("received a state timeout event for different state", "state", data.stateName)
+					return waitForBlock, errTimeoutEventWrongState
+				}
+				// We haven't received the block at current height or height+1. Remove peer.
+				fsm.pool.removePeerAtCurrentHeights(errNoPeerResponse)
+				fsm.resetStateTimer()
+				if len(fsm.pool.peers) == 0 {
+					return waitForPeer, errNoPeerResponseForCurrentHeights
+				}
+				if fsm.pool.reachedMaxHeight() {
+					return finished, nil
+				}
+				return waitForBlock, errNoPeerResponseForCurrentHeights
+
+			case stopFSMEv:
+				if fsm.stateTimer != nil {
+					fsm.stateTimer.Stop()
+				}
+				return finished, errNoErrorFinished
+
+			default:
+				return waitForBlock, errInvalidEvent
+			}
+		},
+	}
+
+	finished = &bReactorFSMState{
+		name: "finished",
+		enter: func(fsm *bReactorFSM) {
+			fsm.logger.Info("Time to switch to consensus reactor!", "height", fsm.pool.height)
+			fsm.toBcR.switchToConsensus()
+			fsm.cleanup()
+		},
+		handle: func(fsm *bReactorFSM, ev bReactorEvent, data bReactorEventData) (*bReactorFSMState, error) {
+			return nil, nil
+		},
+	}
+}
+
+// Interface used by FSM for sending Block and Status requests,
+// informing of peer errors and state timeouts
+// Implemented by BlockchainReactor and tests
+type bcRMessageInterface interface {
+	sendStatusRequest()
+	sendBlockRequest(peerID p2p.ID, height int64) error
+	sendPeerError(err error, peerID p2p.ID)
+	resetStateTimer(name string, timer **time.Timer, timeout time.Duration)
+	switchToConsensus()
+}
+
+func (fsm *bReactorFSM) setLogger(l log.Logger) {
+	fsm.logger = l
+	fsm.pool.setLogger(l)
+}
+
+// Starts the FSM.
+func (fsm *bReactorFSM) start() {
+	_ = fsm.handle(&bcReactorMessage{event: startFSMEv})
+}
+
+// handle processes messages and events sent to the FSM.
+func (fsm *bReactorFSM) handle(msg *bcReactorMessage) error {
+	fsm.mtx.Lock()
```

Comment:
Why do you need lock on fsm?

Reply:
Because processing blocks (CPU intensive) is in a different goroutine.